_posts/code/PerfContest/4/canny-run.sh

```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -o xtrace
+
+wget --no-clobber https://johnysswlab.com/downloads/221575.tar.gz
+tar -xf 221575.tar.gz
+
+CC=clang
+CFLAGS="-O3 -lm -g -ffast-math"
+
+${CC} ${CFLAGS}  canny_source.c -o canny
+
+time ./canny 221575.pgm 0.5 0.7 0.9
```

Comment:
Does the benchmark has validation? If not, then you could as well upload a golden output file and `cmp` it with the output of the benchmark.

Reply:
The thing is, when we are compiling with -ffast-math, we get about 10% better performance because the compiler can vectorize some hot loops. But when compiling with that option, the output file can differ slightly so no binary comparison is possible. I suggest to modify the test script that will build to version of the code, one with fast-math, one without. We use fast-math to measure, but we use nofastmath on the same source code to perform the validation. How does that sound?

Reply:
We use a `-ffast-math` option in a baseline, we do not expect users to remove it since it lowers the performance. So can we just validate against `-ffast-math` output file?
But even then, users could have false-positives if they decide to reorder some FP operations, right? So, maybe use some image diff package like `compare` (https://askubuntu.com/questions/209517/does-diff-exist-for-images) with tolerance threshold (https://imagemagick.org/script/command-line-options.php#similarity-threshold)?

Reply:
Good idea for the comparison. I did the compare on the version with and without fastmath, this is what I get:

```
ivica@ivica-TUF-Gaming-FA706II-FA706II:~/projects/dendibakh.github.io/_posts/code/PerfContest/4$ compare -verbose -metric MAE 221575.pgm_s_0.50_l_0.70_h_0.90.orig.pgm 221575.pgm_s_0.50_l_0.70_h_0.90.pgm test.pgm
221575.pgm_s_0.50_l_0.70_h_0.90.orig.pgm PGM 7680x4320 7680x4320+0+0 8-bit Grayscale Gray 31.6406MiB 1.030u 0:00.190
221575.pgm_s_0.50_l_0.70_h_0.90.pgm PGM 7680x4320 7680x4320+0+0 8-bit Grayscale Gray 31.6406MiB 0.340u 0:00.250
Image: 221575.pgm_s_0.50_l_0.70_h_0.90.orig.pgm
  Channel distortion: MAE
    gray: 0.0197528 (3.01408e-07)
    all: 0.0197528 (3.01408e-07)
221575.pgm_s_0.50_l_0.70_h_0.90.orig.pgm=>test.pgm PGM 7680x4320 7680x4320+0+0 8-bit sRGB 31.6406MiB 4.130u 0:00.859
```

The difference is 0.0197528 for the two versions. When I compare the  processed and unprocessed image, I get the following output:

```
ivica@ivica-TUF-Gaming-FA706II-FA706II:~/projects/dendibakh.github.io/_posts/code/PerfContest/4$ compare -verbose -metric MAE 221575.pgm 221575.pgm_s_0.50_l_0.70_h_0.90.pgm test.pgm
221575.pgm PGM 7680x4320 7680x4320+0+0 8-bit Grayscale Gray 31.6407MiB 1.040u 0:00.199
221575.pgm_s_0.50_l_0.70_h_0.90.pgm PGM 7680x4320 7680x4320+0+0 8-bit Grayscale Gray 31.6406MiB 0.350u 0:00.250
Image: 221575.pgm
  Channel distortion: MAE
    gray: 56405.6 (0.860694)
    all: 56405.6 (0.860694)
221575.pgm=>test.pgm PGM 7680x4320 7680x4320+0+0 8-bit sRGB 31.6407MiB 4.200u 0:00.929
```

Can we say that two images are the same if meam average error is less than 0.3?

Reply:
I think you meant 0.03 (3%), right? I think it should be OK. We need to find a similar tool for Windows,

Reply:
Actually, can we just verify it ourselves? PGM is a simple grayscale format, so I guess we can easily compare it.
Here is a very simple way we go about doing this:
```
template <typename T>
std::vector<T> read_binary_file(const char *fname, size_t num = 0) {
  std::vector<T> vec;
  std::ifstream ifs(fname, std::ios::in | std::ios::binary);
  if (ifs.good()) {
    ifs.unsetf(std::ios::skipws);
    std::streampos file_size;
    ifs.seekg(0, std::ios::end);
    file_size = ifs.tellg();
    ifs.seekg(0, std::ios::beg);
    size_t max_num = file_size / sizeof(T);
    vec.resize(num ? (std::min)(max_num, num) : max_num);
    ifs.read(reinterpret_cast<char *>(vec.data()), vec.size() * sizeof(T));
  }
  return vec;
}

template <typename T>
bool cmp_binary_files(const char *fname1, const char *fname2, T tolerance) {
  const auto vec1 = read_binary_file<T>(fname1);
  const auto vec2 = read_binary_file<T>(fname2);
  if (vec1.size() != vec2.size()) {
    std::cerr << fname1 << " size is " << vec1.size();
    std::cerr << " whereas " << fname2 << " size is " << vec2.size()
              << std::endl;
    return false;
  }
  for (size_t i = 0; i < vec1.size(); i++) {
    if (abs(vec1[i] - vec2[i]) > tolerance) {
      std::cerr << "Mismatch at " << i << ' ';
      if (sizeof(T) == 1) {
        std::cerr << (int)vec1[i] << " vs " << (int)vec2[i] << std::endl;
      } else {
        std::cerr << vec1[i] << " vs " << vec2[i] << std::endl;
      }
      return false;
    }
  }
  return true;
}
```

We can compile it in `validate.exe` and run it after the benchmark. What do you think?
That way we don't have to worry about external packages.